### ISSUE TITLE
ffmpeg doesn't build on Darwin

### ISSUE THREAD

**Issue by henrytill** [NixOS/nixpkgs#7013]
Now that `youtube-dl` depends on `ffmpeg` (#6985), it would be nice if we could try to address the `ffmpeg` situation on Darwin.

Currently it won't build because of a dependency on `SDL`, which isn't supported on Darwin, but that may not be the only problem - it's just the first one currently encountered. 

cc @codyopel: in light of #6602, perhaps you'd like to suggest where/how to address platform-specific dependencies in your generic builder.

**Comment by spwhitt** [comment]
A while back I had ffmpeg building on Darwin, but didn't get it polished enough to be comfortable pushing it out. I've been finalizing and pushing out my ffmepg fixes piecemeal yesterday and today, but in light of #6602 some additional work will clearly be necessary. The fix for the SDL problem you noted is here: https://github.com/NixOS/nixpkgs/pull/7001

I'll push the rest of what I have, and maybe take another crack at this. I'm also interested in feedback from @codyopel, because I have no idea how this null style switching thing is supposed to work.

**Comment by codyopel** [comment]
I didn't have anyway to test for darwin so there was no guarantee it would work as is, but I definitely want to make sure that it does work.

As mentioned by @spwhitt the proper solution for SDL at least is to fix SDL to support darwin.  Beyond SDL however I am almost certain there will be more dependencies that won't build on darwin, but they shouldn't cause build failures.  Normally if a dependency isn't available for a given platform (not supported or just fails to build) it will return null and disable the features that depend on it.  In the case with SDL it is probably returning null, but because there is an assert that requires it for ffplay which is enable by default it fails.  Over all there should be very few changes if any need in ffmpeg directly for cross platform support.  Most changes to get ffmpeg to build are need in ffmpeg's dependencies.

In the meantime overriding ffplay and disabling it may be enough to get ffmpeg to build on darwin.

As for the null switching, it should if the dependency is not available return null and disable the feature.  You can also override any given dependency to be null (same as overriding true to false) which will disable any features that depend on the dependency.  This is very similar to the method of having a variable to enable or disable an optional dependency (e.g. ```variable ? true, dependency ? null```).  However the variable is unnecessary when the dependency is enabled by default and only adds complexity.

**Comment by copumpkin** [comment]
I was just looking at the ffmpeg expression and it's terrifyingly big! Does it really need that many options? I appreciate the effort that went into mirroring the build configurations, but every new option adds exponential blowup to the number of possible states and it's basically impossible to ensure that all combinations work properly. Can't we just optimize for the handful of common cases and make sure it actually works properly on darwin?

What was your goal with making it that configurable?

**Comment by copumpkin** [comment]
> Normally if a dependency isn't available for a given platform (not supported or just fails to build) it will return null

Also, not sure whether you meant it this way, but for any readers who may be confused by it: if something doesn't build on your platform, nix won't magically return null for its derivation. You'll just get an error. Some people have taken additional steps to ensure that certain attributes have null values on certain platforms, but derivations either return a proper completed derivation or fail and return nothing.

**Comment by spwhitt** [comment]
@codyopel 

After fixing the SDL problem, we run into the following while building ffmpeg:

```
error: Package ‘frei0r-plugins-1.4’ in ‘/Users/swhitt/nixpkgs/pkgs/development/libraries/frei0r/default.nix:16’ is not supported on ‘x86_64-darwin’, refusing to evaluate.
For `nixos-rebuild` you can set
  { nixpkgs.config.allowBroken = true; }
in configuration.nix to override this.
For `nix-env` you can add
  { allowBroken = true; }
to ~/.nixpkgs/config.nix.
```

Obviously I could just fix frei0r-plugins on Darwin, but I believe there are a few more interesting observations to make first:

1) If I'm understanding your description of null switching correctly, ffmpeg should simply build with frei0r-plugins turned off. However, this does not happen; instead we see the above evaluation error. It is clear that other dependencies will behave the same way. How should we proceed?
2) I don't believe ffmpeg was dependent on frei0r-plugins prior to the refactor (ie, this dep was disabled by default).
3) frei0r-plugins depends on opencv. Opencv can optionally be compiled with a dependency on ffmpeg (this is extremely common and useful to do). This would yield a circular dependency.

One more question. Consider the following argument to generic.nix:

```
 , libtheora ? null # Theora encoder
```

Callpackage sees the libtheora attribute, so it passes libtheora in. This overrides the null default value. In what situation is the null default value ever used? In other words, how is this ever different from writing this:

```
, libtheora # Theora encoder
```

Thanks!

**Comment by codyopel** [comment]
@copumpkin Keep in mind FFmpeg is not a simple program, but an entire multimedia manipulation framework that is actually fairly complex.  My goal is to have as close as possible to a feature complete ffmpeg package.  Most distros package a very small subset of ffmpeg's dependencies and only Gentoo comes close to having a fully featured build (which is close to the same size if you consider the extra comments I have added to the build which make the Nix version longer).  In the case of Nix there is also the problem of reproducible builds since everyone isn't building ffmpeg locally as with gentoo.  The biggest problem with making it reproducible is that ffmpeg's configure script is a broken nightmare.  It tries to assume to many things at runtime so to counter this I made it so that flags are set explicitly.  If something is not enabled it is explicitly disabled to ensure that build failures are caught.  A number of the optionals are also there to prevent users from disabling codependent features and still allow flexibility.  An example is that you can set configure flags in such a way that even with the ffmpeg executable explicitly enabled it will fail to build the executable, but the build will succeed.  There are also cases where ffmpeg has an internal implementation of a decoder/encoder/filter/etc... and enabling an external library overrides the internal implementation (e.g. using openssl instead of gnutls also requires you to use rtmpdump instead of the internal rtmp implementation which has a slightly different feature set).

Yeah that part about null switching was poorly worded, I'm not sure what I meant.


There are some changes I need to add to all-packages.nix for darwin to disable alsa, pulseaudio, etc..., basically they just need to be set to null.  I need to go through all of the dependencies again to double check what all needs to be disabled on darwin before doing so.

**Comment by spwhitt** [comment]
> My goal is to have as close as possible to a feature complete ffmpeg package.

I just want to say that I support this. I'm a pretty heavy ffmpeg user. It might be nice to have a default ffmpeg package which is compiled light like most distros, then a ffmpeg-heavy for the crazy people like us :).

**Comment by copumpkin** [comment]
@codyopel fully aligned with the overall goal and know the pain of dealing with gnarly packages, but am still not convinced it all needs to be exposed as explicit nix switches. Do you envision people tweaking it much beyond something like what @spwhitt is proposing?

**Comment by henrytill** [comment]
@codyopel: 

@spwhitt and I got `ffmpeg` building over in `##nix-darwin` this evening.

Here's a list of dependencies that need to be nulled to get it building:
```
alsaLib
frei0r
gsm
jack2
libmodplug
libraw1394
libssh
libva
libvpx
openal
opencv
pulseaudio
rtmpdump
samba
v4l_utils
wavpack
xavs
```
For testing, we put this list this in `all-packages.nix` like so:
```
ffmpeg_2_5 = callPackage ../development/libraries/ffmpeg/2.5.nix {
  alsaLib = if stdenv.isDarwin then null else alsaLib;
  (...and so on...)
};
```
Obviously doing this here has the big drawback of needing to do this for every version, but we've only tested this against 2.5.  Do you have any ideas about how to improve upon this?

We also need to add this to `configureFlags` in `ffmpeg`'s `generic.nix`:
```
(if stdenv.isDarwin then "--cc=clang" else null)
```
@spwhitt is putting together PRs to get `fftw` and `soxr` building on Darwin, and with these changes we should have this situation sorted.

**Comment by spwhitt** [comment]
Quick note that I prefer the following:

```
++ optional stdenv.cc.cc.isClang [ "--cc=clang" ];
```

**Comment by henrytill** [comment]
:+1:

**Comment by codyopel** [comment]
@spwhitt
I was discussing with someone else possibly adding a minimal version that consists only of the libraries for programs to build against in addition to the current ffmpeg package, but it wouldn't include any external dependencies beyond those required to build the libraries.  This would also resolve the circular dependency issues you mentioned if only the ffmpeg libraries are needed to build opencv.

And no the null switching doesn't quite work like that, @copumpkin described it properly

Relating to ```libtheora``` vs. ```libtheora ? null```,  ```? null``` allows you to override it to null so that it won't build with that dependency.

@copumpkin
Overall I don't expect most people to change the defaults, beyond enabling features disabled by default.  However are you referring to exposing the optional external dependencies with the null switching or to the configure options like Build/Program/Library options, I suppose both.  As far as the null switching of the optional external dependencies it is used to explicitly enable or disable the configure flag thus the reason for having it (also if darwin overrides are added in all-packages is doesn't require retooling), not just for configurability (and for the issue of external libraries overriding internal features being a use case for disabling them).  For the other sections of the build like Build/Program/Library options, not all of them are necessary, but the library options for example, there is a dependency tree of certain libraries being dependent on one another and also some libraries have external library dependencies, which are used in the asserts (example relating to libraries [here](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/ffmpeg/generic.nix#L236L244)).  Most of this build is to get ffmpeg to build consistently and reproducibly rather than for modifying it, all while still allowing one to do so.

@henrytill 
Thanks for checking what isn't building on darwin, some of those need to be fixed to support darwin.
The libvpx 1.3.0 release unfortunately does not support OSX 10.10, and it isn't building on 10.9.
Alsa, pulseaudio, v4l-utils and a couple others actually need to be disabled on darwin and I will open a PR tonight to start making all the necessary changes that will add the changes for all versions of ffmpeg.

**Comment by henrytill** [comment]
@codyopel: Much appreciated, thanks.  

Just to clarify, it's not just that these packages aren't building or Darwin, or aren't disabled for Darwin by their `meta.platforms` attributes.  If they aren't explicitly set to `null` by the `ffmpeg` expression, you get an error like so:
```
$ nix-build -A ffmpeg
error: Package ‘alsa-lib-1.0.28’ in ‘/Users/ht/src/nixpkgs/pkgs/os-specific/linux/alsa-lib/default.nix:36’ is not supported on ‘x86_64-darwin’, refusing to evaluate.
```
But, yes, some of things can and should eventually get fixed on Darwin.  It would be nice to have `ffmpeg` building for Darwin users before that.

**Comment by codyopel** [comment]
Yes, for right now I am probably going to disable everything on darwin that isn't working and fix the dependencies as I get to them.

It looks like the following we should be able to get working on darwin at some point in the future:
+ frei0r-plugins
+ gsm
+ jack audio
+ libssh
+ libvpx (pic needs to be enabled for darwin as well)
+ openal
+ opencv
+ pulseaudio (not sure if pulseaudio is need on osx)
+ rtmpdump (there might be a problem with the package beacuse ffmpeg fails to build on linux with rtmpdump, it is only enabled(/required) if using openssl)
+ wavpack
+ xavs

The linux only packages are:
+ alsa
+ libraw1394
+ libva
+ v4l-utils

**Comment by henrytill** [comment]
>Quick note that I prefer the following:
> ```
>++ optional stdenv.cc.cc.isClang [ "--cc=clang" ];
>```

@codyopel, update: this won't work on other platforms in this form.  

See https://github.com/NixOS/nixpkgs/pull/7023#issuecomment-87017534 for correct usage.

**Comment by codyopel** [comment]
@henrytill 
Thanks, that is good to know.

@henrytill @spwhitt @copumpkin 
Please let me know if these changes resolve the issues on darwin, if not let me know what needs to be fixed, thanks.
